### ISSUE TITLE
Feature Request: Ability to Pass Authentication Parameters Individually 

### ISSUE THREAD

**Issue by mikaelweave** [Azure/login#39]
It would be great if this action allowed passing in the Subscription ID, Tenant ID, Client ID, and Client Secret as individual parameters instead of bundled in a single JSON object (`creds` today). This would allow for easier use in pipelines that use Terraform (as it requires them split out). It would also make secret rotation easier as you don't have to rebuilt the JSON object.

Potential syntax (as an alternative to the existing syntax):

```yaml
- uses: azure/login@v1.1
      with:
        client_id: ${{ secrets.CLIENT_ID }}
        client_secret: ${{ secrets.CLIENT_SECRET }}
        subscription_id: ${{ secrets.SUBSCRIPTION_ID }}
        tenant_id: ${{ secrets.TENANTID }}
```

If you are open to PRs, I'd love to try this!

**Comment by ericsampson** [comment]
We would really love this too. We have the same 4 separate secret setup as you @mikaelweave , used with Terraform : )
It would be awesome to not need duplicate secrets as JSON to pass into this Action.

**Comment by kmkumaran** [comment]
The reason the login Action took the [credential parameters\(https://github.com/Azure/login/blob/master/README.md#configure-deployment-credentials) as single Json is with the expectation that the getting started experience for users will be simplified with just one copy / paste of command line output as against multiple user actions. 

It appears with Terraform, the need is to have them independent. @N-Usha / @azooinmyluggage / @nadesu - Can you follow-up on this ask and see how we can meet this need and still retain simplified experience for users who would like to opt the current way.

**Comment by ericsampson** [comment]
Thanks @kmkumaran 👍

**Comment by azooinmyluggage** [comment]
One other reason was to use the output of the az ad sp create-for-rbac command as an input. Both arguments at the end are for keeping it simple for end user. 

But looking at the description by @ericsampson and the up votes, this needs to change.

@ericsampson we are open to PR.  How can we solve it gracefully? Two potential guidelines are:
1. Only one set of credentials work. Is there any precedence on one over the other?
2. Do we need any renaming? Having a json based input called "creds" and then four other inputs. Is there anyway we can implement to avoid any confusion for a first day user?

cc @N-Usha

**Comment by ericsampson** [comment]
Here's a proposal that might address all those questions, and is still simple for the new user; these 4 individual inputs would be for more advanced users.

- existing `creds` input works as it does now (except that it's optional)
- add the following 4 new inputs, with the `creds_` prefix to help indicate that they are related to the creds JSON blob:

```
- uses: azure/login@v1.1
      with:
        creds_client_id: ${{ secrets.CLIENT_ID }}
        creds_client_secret: ${{ secrets.CLIENT_SECRET }}
        creds_subscription_id: ${{ secrets.SUBSCRIPTION_ID }}
        creds_tenant_id: ${{ secrets.TENANTID }}
```

- _Key point: any `creds_` input overrides the corresponding value in the `creds` JSON (if passed in by the user)_. This is similar to how ASP.NET Core configuration providers allow settings to be overlayed.
This is actually kind of nice for some use cases, like below.

In other words:
```
- uses: azure/login@v1.1
      with:
        creds: ${{ secrets.AZURE_CREDENTIALS }}
        creds_client_secret: ${{ secrets.CLIENT_SECRET }} # overrides the secret contained in the creds JSON blob.
```
and the basic simple use case still works unchanged:
```
- uses: azure/login@v1.1
      with:
        creds: ${{ secrets.AZURE_CREDENTIALS }}
```

**Comment by azooinmyluggage** [comment]
@N-Usha and @kaverma can you please take a look? Looks good to me.
Thanks @ericsampson

**Comment by kaverma** [comment]
> @N-Usha and @kaverma can you please take a look? Looks good to me.
> Thanks @ericsampson

Thanks, the proposal looks good to me as well.

**Comment by ericsampson** [comment]
Thanks all! 🥳

**Comment by kmkumaran** [comment]
With creds becoming an optional parameter, the Action should check for at-least one of the two inputs given as part of parameter validation.

**Comment by N-Usha** [comment]
Thanks @ericsampson . I liked your idea of supporting both options together... and also creds_ overriding creds if both are provided.
we need to however validate that this new creds_ option covers all the functionality that we support today including support for govt clouds, Azure stack and Support for Az PowerShell. Also today SPN scoping for resource group is also possible using the creds json object... need to consider future extensibility of this solution for all the scopes with the new creds_ option.

@ericsampson - It'll be great if you could also create a PR with the proposed changes.. Please confirm.

**Comment by ericsampson** [comment]
Hi @N-Usha , I won't have time to implement this. I'm already working on other OSS PRs for various repos. Cheers

**Comment by anatbal** [comment]
Hey @BALAGA-GAYATRI, @N-Usha ,
We have come across this issue and considered contributing. Are you currently actively working on it? If you are open to PRs, we are available and we'd love to help with it.

**Comment by anatbal** [comment]
@ericsampson , have you tried this workaround?  This way you don't need to duplicate secrets as JSON and just use your existing vars.
```yaml
   - name: Azure Login
        uses: Azure/login@v1.1
        with:
          creds: '{"clientId":"${{ secrets.ARM_CLIENT_ID }}","clientSecret":"${{ secrets.ARM_CLIENT_SECRET }}","subscriptionId":"${{ secrets.ARM_SUBSCRIPTION_ID }}","tenantId":"${{ secrets.ARM_TENANT_ID }}"}'
```

**Comment by ericsampson** [comment]
@anatbal good idea, I can't recall if we did. 

This doesn't end up printing out the secrets in the build log does it?

Being able to pass them as individual params is easier for people to use, but it's good to have a workaround until then. Cheers :)

**Comment by jmservera** [comment]
I would also prefer to use individual secrets due to this potential issue as noted in the [GitHub Documentation](https://docs.github.com/en/actions/reference/encrypted-secrets):

> To help ensure that GitHub redacts your secret in logs, avoid using structured data as the values of secrets. For example, avoid creating secrets that contain JSON or encoded Git blobs.

Any change in the tooling could potentially leak these secrets. In fact, if you try to "echo" the secret, it is not redacted by the platform. This should be documented as a risk when using the full json as a secret.

**Comment by ericsampson** [comment]
Yeah, I always thought it was odd that this Action is designed to work in a way that GH's own documentation recommends as a security anti-practice.

Similarly wonky is the Azure App Service deploy Action, which uses an XML PublishProfile (containing secrets) and recommends storing it as a GH secret, contrary to GH advisory. 

It would sure be nice if the two companies, being under the same banner, could get their story straight around multiline/structured GH secret usage. It doesn't give a person the warm fuzzies given that it's a security issue.

**Comment by kaverma** [comment]
@BALAGA-GAYATRI , can you please have a look at this PR and see if we can merge and release the same? https://github.com/Azure/login/pull/108

**Comment by bbl** [comment]
> We have added OIDC login mechanism to the action which now accepts individual parameters as inputs. Please refer [sample workflow](https://github.com/Azure/login#sample-workflow-that-uses-azure-login-action-using-oidc-to-run-az-cli-linux) and [configure OIDC](https://github.com/Azure/login#configure-a-service-principal-with-a-federated-credential-to-use-oidc-based-authentication) for more details. So closing this enhancement.

Great news! 
Although, it seems we need to hardcode branch name in the OIDC parameters in azure. It's not an option if you have this pattern in a GitHub workflow:

```yaml
on:
  push:
```

or, e.g.

```yaml
on:
  push:
    - 'dev*' 
```

Is there any option to accept _all_ branches or to use some pattern?
